src/SonarScanner.MSBuild.PreProcessor/JreCaching/JreCache.cs

```diff
@@ -45,6 +47,93 @@ public JreCacheResult IsJreCached(string sonarUserHome, JreDescriptor jreDescrip
         return new JreCacheFailure($"The JRE cache directory in '{Path.Combine(sonarUserHome, "cache")}' could not be created.");
     }
 
+    public async Task<JreCacheResult> DownloadJreAsync(string sonarUserHome, JreDescriptor jreDescriptor, Func<Task<Stream>> jreDownload)
+    {
+        if (EnsureCacheRoot(sonarUserHome, out var cacheRootLocation)
+            && EnsureDirectoryExists(Path.Combine(cacheRootLocation, jreDescriptor.Sha256)) is { } jreDownloadPath)
+        {
+            var downloadTarget = Path.Combine(jreDownloadPath, jreDescriptor.Filename);
+            return fileWrapper.Exists(downloadTarget)
+                ? await UnpackJre(downloadTarget, jreDescriptor, cacheRootLocation)
+                : await DownloadAndUnpackJre(jreDownloadPath, downloadTarget, jreDescriptor, cacheRootLocation, jreDownload);
+        }
+        else
+        {
+            return new JreCacheFailure($"The JRE cache directory in '{Path.Combine(sonarUserHome, "cache", jreDescriptor.Sha256)}' could not be created.");
+        }
+    }
+
+    private async Task<JreCacheResult> DownloadAndUnpackJre(string jreDownloadPath, string downloadTarget, JreDescriptor jreDescriptor, string cacheRootLocation, Func<Task<Stream>> jreDownload)
+    {
+        if (await DownloadJre(jreDownloadPath, downloadTarget, jreDownload) is { } exception)
+        {
+            if (fileWrapper.Exists(downloadTarget)) // Even though the download failed, there is a small chance the file was downloaded by another scanner in the meantime.
+            {
+                return await UnpackJre(downloadTarget, jreDescriptor, cacheRootLocation);
+            }
+            return new JreCacheFailure(string.Format(Resources.ERR_JreDownloadFailed, exception.Message));
+        }
+        else
+        {
+            return await UnpackJre(downloadTarget, jreDescriptor, cacheRootLocation);
+        }
+    }
+
+    private async Task<Exception> DownloadJre(string jreDownloadPath, string downloadTarget, Func<Task<Stream>> jreDownload)
+    {
+        // We download to a temporary file in the right folder.
+        // This avoids conflicts, if multiple scanner try to download to the same file.
+        var tempFileName = Path.GetRandomFileName();
+        var tempFile = Path.Combine(jreDownloadPath, tempFileName);
+        try
```

Comment:
The `finally` runs after the catch, so the FileStream is still open. If the FileStream is open, it can not be deleted (File in use). That is why I try to close the stream before deleting the file. I'm not sure if delete would work in your case.

Reply:
I added `Download_DownloadFileNew_Failure_WithTestFiles` in 0f25256d4228652133a4b9edddd9d2fc75ba7245. If we remove the close of the stream, the test fails and the temp file is still in the directory.